File: Olympiad/COCI/official/2008/contest6_solutions/granica.cpp
#include <cstdio>

int n;
int a[100];

int ok( int M ) {
   int ostatak = a[0]%M;
   for( int i = 1; i < n; ++i )
      if( a[i]%M != ostatak )
         return 0;
   return 1;
}

int main( void ) {
   scanf( "%d", &n );
   for( int i = 0; i < n; ++i ) scanf( "%d", &a[i] );

   int razlika = a[0]-a[1];
   if( razlika < 0 ) razlika = -razlika;

   for( int d = 1; d*d <= razlika; ++d ) {
      if( razlika % d != 0 ) continue;

      int M1 = d, M2 = razlika/d;

      if( ok( M1 ) && M1 != 1  ) printf( "%d ", M1 );
      if( ok( M2 ) && M2 != M1 ) printf( "%d ", M2 );
   }
   printf( "\n" );

   return 0;
}

File: Olympiad/COCI/official/2008/contest6_solutions/granica.pas
program granica;

var
   n, i, d, M1, M2, razlika : longint;
   a : array[1..100] of longint;

function ok( M : longint ) : boolean;
var
   i, ostatak : longint;
begin
   ostatak := a[1] mod M;
   ok := true;
   for i := 1 to n do
      if a[i] mod M <> ostatak then
         ok := false;
end;

begin
   readln( n );
   for i := 1 to n do readln( a[i] );

   razlika := abs(a[1]-a[2]);

   d := 1;
   while d*d <= razlika do begin
      if razlika mod d = 0 then begin
         M1 := d;
         M2 := razlika div d;

         if ok( M1 ) and (M1 <> 1) then write( M1, ' ' );
         if ok( M2 ) and (M2 <> M1) then write( M2, ' ' );
      end;
      d := d + 1;
   end;
   writeln;
end.
